Docs/BazelForiOSDevelopers.md

```diff
@@ -0,0 +1,370 @@
+objc_library(name="some", copts=["-DSOME"])
+```
+
+The same principals can be applied to many rules: wrapping macros with macros.
+With Starlark, the possibilities are endless! _technically they are endless as
+the language is not turing complete_. Please see the
+[`objc_library`](https://docs.bazel.build/versions/master/be/objective-c.html)
+documentation for all possible arguments. _Note: Unlike Starlark rules which
+are easily added on to Bazel, the `objc_library` is part of the internal java
+rules shipped with the Bazel
+binary._ 
+
+### Toolchains
+
+In additional to configuring the build with rules, bazel provides a additional
+primitive, the toolchain. Toolchains provide default compilers and arguments
+for those compilers. For the native c++ rules, this is relied upon to configure
+the many flags required for cross compilation.
+
+For most iOS projects, configuring the c++ toolchain isn't required. Defining a
+custom toolchain is useful to gives full control over compiler invocations or
+to customize the compiler being invoked.  Please see the [toolchain
+documentation](https://docs.bazel.build/versions/master/tutorial/cc-toolchain-config.html)
+to learn more.
+
+### Generated Xcode projects
+
+In an Xcode world, folks checked in a project which contains an listing of
+files, build settings, and IDE state. When the project scales, the project file
+model breaks down. Auditing and code reviewing config changes in Xcode projects
+quickly becomes difficult. With Bazel, human readable `BUILD` files are the
+source of truth for build and Xcode configuration. Tools like
+[XCHammer](https://github.com/pinterest/xchammer), and
+[Tulsi](https://github.com/bazelbuild/tulsi) use an aspect to traverse the
+build graph and extract metadata required to generate a project.  These tools
+make it easier to manage the project and generate on demand - not needing to
+check it in.
+
+XCHammer provides a rule to [bazel build Xcode
+projects](ihttps://github.com/pinterest/xchammer#bazel-build-xcode-projects).
+Simply declare the rule with the project and Bazel build the target.
+```
+load("@xchammer_resources//:xcodeproject.bzl", "xcode_project")
+xcode_project(
+    name = "MyProject",
+    targets = [ "//ios-app:ios-app" ],
+    paths = [ "**" ],
+)
+```
+
+Like any other Bazel target, it's built from the command line
+```
+bazel build :MyProject
+```
+
+The rule definition for the `xcode_project` may be found in the [github
+repository](https://github.com/pinterest/xchammer/blob/master/BazelExtensions/xcodeproject.bzl).
+Simply put, the aspect traverses sources, and invokes the `xchammer` binary with
+a JSON file. Internally, XCHammer instantiates
+[`XcodeGen`](https://github.com/yonaskolb/XcodeGen) types and writes them out to
+disk with `xcodeproj`.
+
+The workflow for generating an Xcode project with XCHammer is quite simple:
+```
+command line -> bazel -> rule -> xchammer -> xcodegen -> xcodeproj
```

Comment:
is vanilla bazel:  bazel -> rule -> xcodeproj?

Reply:
There's no way to generate an Xcode project with Bazel its self.

Reply:
With https://tulsi.bazel.build/ for sure, which is also maintained by Bazel?

Reply:
Tulsi and XCHammer are both external-to-bazel aspects and rules. XCHammer uses tulsi intenrally, but it's an implementation detail